### ISSUE TITLE
Render actual text rather than paths in SVG (possibly with embedded fonts)

### ISSUE THREAD

**Issue by ongun-kanat** [typst/typst#4702]
### Description

While rendering text and fonts as paths in SVG images guarantee a consistent output, it makes them unselectable and unsearchable. The rendered path is also limited by the capability of the rendering engine and the font libraries Typst uses which are currently not as complete as the text rendering infrastructure that come with the OS and browsers.

An option in the typst CLI to opt for SVG `<text>` element output for the textual data would be very useful, especially when the file is to be viewed in the browser.

An optional improvement over this is embedding the actual fonts within the SVG file. This has a potential to bloat the SVG file, but could be useful for things like figures.

### Use Case

I'm using Typst to generate figures containing shapes and tables. I would prefer the text in them to be selectable when viewed in a browser. 

SVG files that use `<text>` elements are also smaller than the SVG files whose text is converted into paths.

**Comment by Enivex** [comment]
Someone more knowledgeable can correct me if I'm wrong, but I do not believe it's possible to embed fonts in an SVG, at least not in a way that's usable outside of a browser. Therefore such SVGs would be non-portable.

(Just a wild thought: Would it be possible to put a transparent text layer on top, just for selectability purposes?)

**Comment by ongun-kanat** [comment]
Yeah it is true that the embedding is done via CSS styling. There are even tools that analyze the local fonts and embed the TTF / WOFF files inside SVGs:

- https://github.com/BTBurke/svg-embed-font
- https://github.com/phauer/svg-buddy

Displaying those fonts correctly of course requires a CSS capable renderer. However, `<style>` tag is a part of the SVG standard (optional for 1.1, mandatory for 2.0): https://svgwg.org/svg2-draft/styling.html#StyleElement . The level of implementation varies quite a bit of course.

I am okay with the invisible text layer as well. As I said it is optional to embed fonts.

Just outputting the SVG with the text tag and hoping the user's system has the font is also an option but I can see that it can cause some unwanted results when the font is fed via CLI.

**Comment by Enivex** [comment]
SVG 2.0 isn't finalized (and I'm not entirely convinced it ever will be)

**Comment by LaurenzV** [comment]
The main problem is that exporting the text as strings in `<text>` elements would only work for the very simplest texts, but Typst supports a lot more (think subscripts, superscripts, specifying OpenType feature to use when shaping the text, and also math mode which (I think) has very complex positioning rules), and I don't think you can reliably represent those in SVG. 

Yes, one option could be to just overlay invisible text on top, but that feels very hacky and I'm not sure how well it would work for more complex scenarios.

**Comment by Bequen** [comment]
> you might like read this -- https://vecta.io/blog/using-fonts-in-svg

Thanks, I did not realize that the `<style>` with `@font-face` needs to be directly inside the `<svg>` tag for `font-name` attribute to work. Unfortunately I cannot add `@font-face` during export to the `<svg>`, because I do not know `src` attribute. For my tests on the HTML page I had `@font-face` defined in the outside css file, but I looks like it do not get picked up by embedded SVG.

Anyway I noticed even Inkscape exports SVG files without fonts embedded and is using `"style: font-family: 'SomeFont'", which looks much better and simpler than using classes. I tried it and it works even for font faces defined outside of the SVG.

**Comment by terefang** [comment]
> Thanks, I did not realize that the `<style>` with `@font-face` needs to be directly inside the `<svg>` tag for `font-name` attribute to work. Unfortunately I cannot add `@font-face` during export to the `<svg>`, because I do not know the value for `src` attribute. For my tests on the HTML page I had `@font-face` defined in the outside css file, but I looks like it do not get picked up by embedded SVG.

you can embed fonts via `data:` urls.

**Comment by Bequen** [comment]
> > Thanks, I did not realize that the `<style>` with `@font-face` needs to be directly inside the `<svg>` tag for `font-name` attribute to work. Unfortunately I cannot add `@font-face` during export to the `<svg>`, because I do not know the value for `src` attribute. For my tests on the HTML page I had `@font-face` defined in the outside css file, but I looks like it do not get picked up by embedded SVG.
> 
> you can embed fonts via `data:` urls.

Good idea. I tried that, but now the SVG is enormous (surprise). The HTML page with several of them, each with redundant fonts, is even bigger. But I think this might be the correct solution:
- portable
- uses pure text (what OP wanted)

The enormous size can be optimized afterwards, by substituting the fonts. But that seems like a very specific optimization, that requires configuring.

But on the other hand, the SVG without embedded fonts can easily be made portable. And there seems to be some tools for that. Or they can open it in Inkscape, convert everything to path, and save.

My point is, if you want to use it for figures on HTML page, it can

**Comment by Bequen** [comment]
> > Thanks, I did not realize that the `<style>` with `@font-face` needs to be directly inside the `<svg>` tag for `font-name` attribute to work. Unfortunately I cannot add `@font-face` during export to the `<svg>`, because I do not know the value for `src` attribute. For my tests on the HTML page I had `@font-face` defined in the outside css file, but I looks like it do not get picked up by embedded SVG.
> 
> you can embed fonts via `data:` urls.

Good idea! I tried that, but now the SVG is enormous. But it might be the correct solution:
- it is portable
- it uses pure, selectable text

On the other hand, optimizing away these embedded fonts seems overly complicated and would even require some configuring. But for using it on the web, this optimization is a must while being quite inconvinient.

So I propose to NOT embed the fonts by default and have a CLI option to enable it. Just like in Clang you can turn on and off some optimizations and warnings. What do you think?

My argument is, that most people are going to use the SVGs for figures on HTML page. Embedding by default would require an extra step for most people. And as I said, not even Inkscape embeds the fonts.

Is there some other reason for portability of SVG specifically that I am missing? And if so, would it be that common?

**Comment by terefang** [comment]
maybe it may be good t also apply the same subsetting techniques to the font embedded in SVGs as in PDFs.

that can reduce the size of large fonts significantly based on the number of characters used.

**Comment by Bequen** [comment]
> maybe it may be good t also apply the same subsetting techniques to the font embedded in SVGs as in PDFs.
> 
> that can reduce the size of large fonts significantly based on the number of characters used.

Unfortunately the method used for PDFs is PDF specific. There are some libraries that can do more general subsetting like
- AllSorts: https://github.com/yeslogic/allsorts
- Klippa: https://github.com/googlefonts/fontations/tree/main/klippa

Klippa is smaller and more focused on subsetting. Unfortunately I cannot even find it as a crate.
